packages/entity-server/src/routes/hierarchy/EntityRelationsRoute.ts

```diff
@@ -0,0 +1,150 @@
+      ancestorType,
+    );
+
+    return Object.entries(descendantAncestorMapping).map(([descendant, { code: ancestor }]) => ({
+      descendant,
+      ancestor,
+    }));
+  }
+
+  async buildFormattedEntitiesByCode(ancestors: EntityType[], descendants: EntityType[]) {
+    const allEntities = [...ancestors, ...descendants];
+    const formattedEntities = await this.req.ctx.formatEntitiesForResponse(allEntities);
+    const formattedEntitiesByCode: Record<string, EntityResponseObject> = {};
+    allEntities.forEach((unformattedEntity, index) => {
+      formattedEntitiesByCode[unformattedEntity.code] = formattedEntities[index];
+    });
+    return formattedEntitiesByCode;
+  }
+
+  async groupByAncestorResponse(
+    descendants: EntityType[],
+    ancestorType: string,
+    flattenToCode: boolean,
+  ) {
+    const { entity, hierarchyId, allowedCountries } = this.req.ctx;
+
+    const ancestorCodeToDescendantCodes = reduceToArrayDictionary(
+      await this.buildDescendantAncestorPairs(descendants, ancestorType),
+      'ancestor',
+      'descendant',
+    );
+
+    if (flattenToCode) {
+      return ancestorCodeToDescendantCodes;
+    }
+
+    const ancestorsWithDescendantsCodes = Object.keys(ancestorCodeToDescendantCodes);
+    const ancestors =
+      entity.type === ancestorType
+        ? [entity]
+        : await entity.getDescendants(hierarchyId, {
+            code: ancestorsWithDescendantsCodes,
+            country_code: allowedCountries,
+          });
+
+    const formattedEntitiesByCode = await this.buildFormattedEntitiesByCode(ancestors, descendants);
+    return ancestors.map(ancestor => ({
+      ...formattedEntitiesByCode[ancestor.code],
+      descendants: ancestorCodeToDescendantCodes[ancestor.code].map(
+        descendantCode => formattedEntitiesByCode[descendantCode],
+      ),
+    }));
+  }
+
+  async groupByDescendantResponse(
+    descendants: EntityType[],
+    ancestorType: string,
+    flattenToCode: boolean,
+  ) {
+    const { entity, hierarchyId, allowedCountries } = this.req.ctx;
+
+    const descendantCodeToAncestorCode = reduceToDictionary(
+      await this.buildDescendantAncestorPairs(descendants, ancestorType),
+      'descendant',
+      'ancestor',
+    );
+
+    if (flattenToCode) {
+      return descendantCodeToAncestorCode;
+    }
+
+    const ancestorsWithDescendantsCodes = Object.values(descendantCodeToAncestorCode);
+    const ancestors =
+      entity.type === ancestorType
+        ? [entity]
+        : await entity.getDescendants(hierarchyId, {
+            code: ancestorsWithDescendantsCodes,
+            country_code: allowedCountries,
+          });
+
+    const formattedEntitiesByCode = await this.buildFormattedEntitiesByCode(ancestors, descendants);
+    return descendants.map(descendant => ({
+      ...formattedEntitiesByCode[descendant.code],
+      ancestor: formattedEntitiesByCode[descendantCodeToAncestorCode[descendant.code]],
+    }));
+  }
+
+  async buildResponse() {
```

Comment:
I find it convenient to declare class methods top/down: the base calling method first, then the ones it calls etc. Based on this, `buildResponse` would be placed on the top of this class

Reply:
This is unresolved but not a big deal - a detail really.

Reply:
Funnily I actually got a contradicting recommendation from @edmofro that we prefer to put all functions called by another function above that function (admittedly that was not for class but a standard script file). I don't think I have too much of a preference either way 🙂

Reply:
I have discussed this with Edwin in the past, and I believe the conclusion was: top-down structure for classes, bottom-up for functions, right @edmofro ?

Perhaps worth quickly mentioning in dev-chat to get it across the team

Reply:
That's right @kael89!

And yes, I think worth bringing up at dev chat